libfreerdp/core/transport.c

```diff
@@ -1809,3 +1870,191 @@ void transport_set_early_user_auth_mode(rdpTransport* transport, BOOL EUAMode)
+{
+	rdpTransportLayer* layer =
+	    (rdpTransportLayer*)calloc(1, sizeof(rdpTransportLayer) + contextSize);
+	if (!layer)
+		return NULL;
+
+	if (contextSize)
+		layer->userContext = layer + 1;
+
+	return layer;
+}
+
+void transport_layer_free(rdpTransportLayer* layer)
+{
+	if (!layer)
+		return;
+
+	IFCALL(layer->Close, layer->userContext);
+	free(layer);
+}
+
+static int transport_layer_bio_write(BIO* bio, const char* buf, int size)
+{
+	if (!buf || !size)
+		return 0;
+	if (size < 0)
+		return -1;
+
+	rdpTransportLayer* layer = (rdpTransportLayer*)BIO_get_data(bio);
+	if (!layer)
+		return -1;
+
+	BIO_clear_flags(bio, BIO_FLAGS_WRITE | BIO_FLAGS_SHOULD_RETRY);
+
+	int status = IFCALLRESULT(-1, layer->Write, layer->userContext, buf, size);
+
+	if (status >= 0 && status < size)
+		BIO_set_flags(bio, (BIO_FLAGS_WRITE | BIO_FLAGS_SHOULD_RETRY));
+
+	return status;
+}
+
+static int transport_layer_bio_read(BIO* bio, char* buf, int size)
+{
+	if (!buf || !size)
+		return 0;
+	if (size < 0)
+		return -1;
+
+	rdpTransportLayer* layer = (rdpTransportLayer*)BIO_get_data(bio);
+	if (!layer)
+		return -1;
+
+	BIO_clear_flags(bio, BIO_FLAGS_READ | BIO_FLAGS_SHOULD_RETRY);
+
+	int status = IFCALLRESULT(-1, layer->Read, layer->userContext, buf, size);
+
+	if (status == 0)
+		BIO_set_flags(bio, (BIO_FLAGS_READ | BIO_FLAGS_SHOULD_RETRY));
+
+	return status;
+}
+
+static int transport_layer_bio_puts(BIO* bio, const char* str)
+{
+	return 1;
+}
+
+static int transport_layer_bio_gets(BIO* bio, char* str, int size)
+{
+	return 1;
+}
+
+static long transport_layer_bio_ctrl(BIO* bio, int cmd, long arg1, void* arg2)
+{
+	rdpTransportLayer* layer = (rdpTransportLayer*)BIO_get_data(bio);
+	if (!layer)
+		return -1;
+
+	int status = -1;
+	switch (cmd)
+	{
+		case BIO_C_GET_EVENT:
+			*((HANDLE*)arg2) = IFCALLRESULT(NULL, layer->GetEvent, layer->userContext);
+			status = 1;
+			break;
+
+		case BIO_C_SET_NONBLOCK:
+			status = 1;
+			break;
+
+		case BIO_C_WAIT_READ:
+		{
+			int timeout = (int)arg1;
+			BOOL r = IFCALLRESULT(FALSE, layer->Wait, layer->userContext, FALSE, timeout);
+			/* Convert timeout to error return */
+			if (!r)
+			{
+				errno = ETIMEDOUT;
+				status = 0;
+			}
+			else
+				status = 1;
+			break;
+		}
+
+		case BIO_C_WAIT_WRITE:
+		{
+			int timeout = (int)arg1;
+			BOOL r = IFCALLRESULT(FALSE, layer->Wait, layer->userContext, TRUE, timeout);
+			/* Convert timeout to error return */
+			if (!r)
+			{
+				errno = ETIMEDOUT;
+				status = 0;
+			}
+			else
+				status = 1;
+			break;
+		}
+
+		case BIO_CTRL_GET_CLOSE:
+			status = BIO_get_shutdown(bio);
+			break;
+
+		case BIO_CTRL_SET_CLOSE:
+			BIO_set_shutdown(bio, (int)arg1);
+			status = 1;
+			break;
+
+		case BIO_CTRL_FLUSH:
+		case BIO_CTRL_DUP:
+			status = 1;
+			break;
+
+		default:
+			status = 0;
+			break;
+	}
+
+	return status;
+}
+
+static int transport_layer_bio_new(BIO* bio)
+{
+	BIO_set_flags(bio, BIO_FLAGS_SHOULD_RETRY);
```

Comment:
`WINPR_ASSERT` before dereferencing or return `error` after a failed check (spot unexpected arguments early on, debugging in the ssl layers is hard)